code/modules/food_and_drink/cakes.dm

```diff
@@ -24,59 +24,445 @@
+
+			var/transferamount = (src.amount/src.clayer)/slices //amount of reagent to transfer to slices
+			for(var/i=1,i<=slices,i++) //generating child slices of the parent template
+				var/obj/item/reagent_containers/food/snacks/cake/custom/schild = new /obj/item/reagent_containers/food/snacks/cake/custom
+				schild.icon_state = "slice-overlay"
+				for(var/i2=1,i2<=s.overlays.len,i2++) //looping through parent overlays and copying them over to the children
+					schild.UpdateOverlays(s.GetOverlayImage("[s.overlay_refs[i2]]"),"[s.overlay_refs[i2]]")
+				if(candle) //making sure there's only one candle :)
+					if(candle == 1)
+						schild.UpdateOverlays(new /image('icons/obj/foodNdrink/food_dessert.dmi',"slice-candle"),"slice-candle")
+					if(candle == 2)
+						schild.UpdateOverlays(new /image('icons/obj/foodNdrink/food_dessert.dmi',"slice-candle_lit"),"slice-candle_lit")
+					candle = 0
+					if(src.litfam) //light update
+						src.put_out()
+						schild.light()
+				src.reagents.trans_to(schild,transferamount) //setting up the other properties of the slice
+				schild.pixel_x = rand(-6, 6)
+				schild.pixel_y = rand(-6, 6)
+				for(var/b=1,b<=src.food_effects.len,b++)
+					if(src.food_effects[b] in schild.food_effects)
+						continue
+					schild.food_effects += src.food_effects[b]
+				schild.w_class = 1
+				schild.quality = src.quality
+				schild.name = "slice of [src.name]"
+				schild.desc = "a delicious slice of cake!"
+				schild.food_color = src.food_color
+				schild.sliced = 1
+				schild.amount = 1
+
+				schild.set_loc(get_turf(src.loc))
+				/*if(src in user.contents) //spawning
+					schild.set_loc(get_turf(user))
+				else
+					schild.set_loc(get_turf(src))*/
+			qdel(s) //cleaning up the template slice
+			if(src.clayer == 1) //qdel(src) if there was only one layer to the cake, otherwise, decrement the layer
+				qdel(src)
+			else
+				src.clayer--
+			return
+		else if((!(src.litfam)) && (((istype(W,/obj/item/weldingtool) && W:welding)) || (istype(W, /obj/item/device/igniter))) || \
+		(((istype(W, /obj/item/match)) || (istype(W, /obj/item/device/light/candle)) || (istype(W, /obj/item/clothing/head/cakehat)) || (istype(W, /obj/item/device/light/zippo))) && (W:on))) //we don't talk about this......
+			src.light() //so basically...if the cake doesnt have a light source AND ([the weapon is a welding tool and on, or an igniter] OR [its a bunch of other things AND those things are on])
+
+		switch(W.type) //this is where the mess starts...
+			if(/obj/item/reagent_containers/food/drinks/drinkingglass/icing) //handling for frosting overlays. they have a color, so theyre handled a bit differently
+				if(!(W.reagents.total_volume >= 25))
+					boutput(user, "<span style=\"color:red\">The icing tube isn't full enough to frost the cake!</span>")
+					return
+				var/frostingtype
+				frostingtype = input("Which frosting style would you like?", "Frosting Style", null) as null|anything in frostingstyles
```

Comment:
It would be nice and much more user-friendly for mass-making of cakes if this selection could be transferred over to the icing tube self attack.